svc.go

```diff
@@ -0,0 +1,244 @@
+	w := ggio.NewDelimitedWriter(s)
+
+	for {
+		var req pb.Message
+		var res pb.Message
+
+		err := r.ReadMsg(&req)
+		if err != nil {
+			s.Reset()
+			return
+		}
+
+		t := req.GetType()
+		switch t {
+		case pb.Message_REGISTER:
+			r := rz.handleRegister(pid, req.GetRegister())
+			res.Type = pb.Message_REGISTER_RESPONSE.Enum()
+			res.RegisterResponse = r
+			err = w.WriteMsg(&res)
+			if err != nil {
+				log.Debugf("Error writing response: %s", err.Error())
+				s.Reset()
+				return
+			}
+
+		case pb.Message_UNREGISTER:
+			err := rz.handleUnregister(pid, req.GetUnregister())
+			if err != nil {
+				log.Debugf("Error unregistering peer: %s", err.Error())
+			}
+
+		case pb.Message_DISCOVER:
+			r := rz.handleDiscover(pid, req.GetDiscover())
+			res.Type = pb.Message_DISCOVER_RESPONSE.Enum()
+			res.DiscoverResponse = r
+			err = w.WriteMsg(&res)
+			if err != nil {
+				log.Debugf("Error writing response: %s", err.Error())
+				s.Reset()
+				return
+			}
+
+		default:
+			log.Debugf("Unexpected message: %s", t.String())
+			s.Reset()
+			return
+		}
+	}
+}
+
+func (rz *RendezvousService) handleRegister(p peer.ID, m *pb.Message_Register) *pb.Message_RegisterResponse {
+	ns := m.GetNs()
+	if ns == "" {
+		return newRegisterResponseError(pb.Message_E_INVALID_NAMESPACE, "unspecified namespace")
+	}
+
+	if len(ns) > MaxNamespaceLength {
+		return newRegisterResponseError(pb.Message_E_INVALID_NAMESPACE, "namespace too long")
+	}
+
+	mpi := m.GetPeer()
+	if mpi == nil {
+		return newRegisterResponseError(pb.Message_E_INVALID_PEER_INFO, "missing peer info")
+	}
+
+	mpid := mpi.GetId()
+	if mpid != nil {
+		mp, err := peer.IDFromBytes(mpid)
+		if err != nil {
+			return newRegisterResponseError(pb.Message_E_INVALID_PEER_INFO, "bad peer id")
+		}
+
+		if mp != p {
+			return newRegisterResponseError(pb.Message_E_INVALID_PEER_INFO, "peer id mismatch")
+		}
+	}
+
+	maddrs := mpi.GetAddrs()
+	if len(maddrs) == 0 {
+		return newRegisterResponseError(pb.Message_E_INVALID_PEER_INFO, "missing peer addresses")
+	}
+
+	mlen := 0
+	for _, maddr := range maddrs {
+		mlen += len(maddr)
+	}
+	if mlen > MaxPeerAddressLength {
+		return newRegisterResponseError(pb.Message_E_INVALID_PEER_INFO, "peer info too long")
+	}
+
+	// Note:
+	// We don't validate the addresses, because they could include protocols we don't understand
+	// Perhaps we should though.
+
+	mttl := m.GetTtl()
+	if mttl < 0 || mttl > MaxTTL {
+		return newRegisterResponseError(pb.Message_E_INVALID_TTL, "bad ttl")
+	}
+
+	ttl := DefaultTTL
+	if mttl > 0 {
+		ttl = int(mttl)
+	}
+
+	// now check how many registrations we have for this peer -- simple limit to defend
+	// against trivial DoS attacks (eg a peer connects and keeps registering until it
+	// fills our db)
+	rcount, err := rz.DB.CountRegistrations(p)
+	if err != nil {
+		log.Errorf("Error counting registrations: %s", err.Error())
+		return newRegisterResponseError(pb.Message_E_INTERNAL_ERROR, err.Error())
```

Comment:
including raw text from a sql error in a response to the client seems like a potentially bad information leak. Just a gut feeling though.

Reply:
Plugged the leaks in https://github.com/libp2p/go-libp2p-rendezvous/pull/1/commits/baf1e4e61812c08f5e25ac019e04cbabdced4ce4

It's just a "database error" to the client now.